### ISSUE TITLE
Too many errors reported 

### ISSUE THREAD

**Issue by leppie** [dotnet/roslyn#7177]
For the following:
```
var s = typeof(Action<0>);
```
7 errors gets reported
```
(4,30,4,31): Error CS1031: Type expected
(4,30,4,31): Error CS1003: Syntax error, '>' expected
(4,30,4,31): Error CS1026: ) expected
(4,30,4,31): Error CS1002: ; expected
(4,32,4,33): Error CS1525: Invalid expression term ')'
(4,32,4,33): Error CS1002: ; expected
(4,32,4,33): Error CS1513: } expected
```
Only the 1st two are informative. Seems parser recovery is not catering here.

Along similar lines:
```
var s = new Action<0>();
```
6 errors:
```
(4,27,4,28): Error CS1031: Type expected
(4,27,4,28): Error CS1003: Syntax error, '>' expected
(4,27,4,28): Error CS1526: A new expression requires (), [], or {} after type
(4,27,4,28): Error CS1002: ; expected
(4,30,4,31): Error CS1525: Invalid expression term ')'
(4,16,4,27): Error CS1729: 'Action<?>' does not contain a constructor that takes 0 arguments
```
Another example:
```
Action<static> a = null;
```
Errors:
```
(3,15,3,21): Error CS1525: Invalid expression term 'static'
(3,15,3,21): Error CS1002: ; expected
(3,15,3,21): Error CS0106: The modifier 'static' is not valid for this item
(3,21,3,22): Error CS1031: Type expected
(3,21,3,22): Error CS1001: Identifier expected
(3,21,3,22): Error CS1002: ; expected
(3,21,3,22): Error CS1525: Invalid expression term '>'
```

**Comment by CnSimonChan** [comment]
I think any obvious syntax problem will make Roslyn crazy and report several errors.

    // Oh I misplaced a `static`!
    return AsyncInfo.Run<int>((cancelToken, progress) => static Task.FromResult(true));

Roslyn: This code is a piece of s**t!!!

````
(96,66,96,72): error CS1525: Invalid expression term 'static'
(96,66,96,72): error CS1026: ) expected
(96,66,96,72): error CS1002: ; expected
(96,66,96,72): error CS0106: The modifier 'static' is not valid for this item
(96,88,96,89): error CS1001: Identifier expected
(96,88,96,94): error CS1528: Expected ; or = (cannot specify constructor arguments in declaration)
(96,88,96,89): error CS1003: Syntax error, '[' expected
(96,94,96,95): error CS1003: Syntax error, ']' expected
(96,94,96,95): error CS1002: ; expected
(96,94,96,95): error CS1513: } expected
````

**Comment by CyrusNajmabadi** [comment]
You're hitting the general problem in parsers of trying to decide if an errant piece of syntax should be ignored, or if it should start some other construct.  When the parser sees "static Task" it knows it isn't value in the expression context of a lambda.  However "static Task" very much looks like it may be starting a new member.  Consider the far more likely scenario of:

```c#
AsyncInfo.Run<int>((cancelToken, progress) => 

static Task SomeFunction()
```

Clearly in this case we want to stop parsing out the lambda (and it's containing statement/member), and we want to start parsing out a new method.

So the parser makes judicious decisions when it sees bad code about what to do.  Sometimes it will ignore what it seems and will keep on going.  Sometimes it will stop what it is doing and allow another parsing function to take over.

Fortunately, all of these are heuristics.  And as the compiler is hand-written (and not auto-generated), it's not too difficult to improve the error handling in these cases.  So feel free in both these cases to just improve the parser to do better here.

**Comment by leppie** [comment]
Another one, but this could probably be fixed easier/separately:
```
using System;

enum Foo { Bar, Baz }

public class P
{
    static void Foo(Foo foo = Foo.Baz) {}

    static void Main()
    {
        Foo();
    }
}
```
Reports:
```
error CS0119: 'P.Foo(Foo)' is a method, which is not valid in the given context
error CS1750: A value of type '?' cannot be used as a default parameter because there are no standard conversions to type 'Foo'
```
The first is correct. But the second is non-sensical due to the type being an error type already. I dont think this should be reported.

**Comment by gafter** [comment]
Grabbed by @rfairfax

**Comment by gafter** [comment]
Fixed in #7624
